Fix a bug when parallel commit is on

YKatsuragi opened the pull request:
This commit fixes the timing of calling pgfdw_reset_xact_state when parallel commit is on.

Previously pgfdw_reset_xact_state was called before pgfdw_finish_commit_prepared_cleanup when parallel commit was on. Therefore, connections were not usable in pgfdw_finish_commit_prepared_cleanup and warning was raised.

Reported-by: Etsuro Fujita
Author: Etsuro Fujita

MasaoFujii commented:
@YKatsuragi 
Thanks for the report and patch! The proposed patch looks good to me.

Additionally, in order to ensure that the reported issue has been resolved, it may be worthwhile to include a test case for parallel_commit=on in the regression test. One way to do this would be to specify the option "parallel_commit 'on'" for the pgfdw_plus_loopback2 server in postgres_fdw_plus.sql regression test file. Thought?

YKatsuragi commented:
@MasaoFujii 
Thank you for the comment!

> in order to ensure that the reported issue has been resolved, it may be worthwhile to include a test case for parallel_commit=on in the regression test.

Yes.

> One way to do this would be to specify the option "parallel_commit 'on'" for the pgfdw_plus_loopback2 server in postgres_fdw_plus.sql regression test file.

I think all the existing test cases should be passed even when parallel_commit=on, so I add the parallel_commit option to CREATE SERVER statement.
However, adding `parallel_commit 'on'` to loopback2 caused the following test failure.
```
@@ -538,19 +538,19 @@
 SAVEPOINT s;
 INSERT INTO ft1 VALUES (320);
 INSERT INTO ft2 VALUES (320);
+ERROR:  SAVEPOINT can only be used in transaction blocks
 COMMIT;
 SELECT split_part(query, '_', 1) FROM pg_stat_activity
     WHERE application_name LIKE 'pgfdw_plus_loopback%' ORDER BY query;
-         split_part
-----------------------------
- PREPARE TRANSACTION 'pgfdw
- PREPARE TRANSACTION 'pgfdw
-(2 rows)
+   split_part
+----------------
+ DEALLOCATE ALL
+(1 row)

 SELECT count(*) FROM pgfdw_plus_resolve_foreign_prepared_xacts_all();
  count
 -------
-     2
+     0
 (1 row)

 -- ===================================================================
@@ -668,7 +668,7 @@
 DETAIL:  Error message: password is required
  count
 -------
-     3
+     2
 (1 row)

 -- ===================================================================
```


I think this is because the ConnCacheEntry used for loopback2 is not reset after a transaction.
In this test case, we set parallel_commit and pgfdw_skip_commit_phase to on, so the connection cache entry is not appended to pending list in pgfdw_commit_prepared() and the entry is not reset in pgfdw_finish_commit_prepared_cleanup().
Because of this, a new transaction uses the not cleaned up connection entry (entry->xact_depth = 1) to make a savepoint (with no BEGIN).

To fix this, I added codes to reset connection cache entry in pgfdw_commit_prepared().
What do you think about this?

YKatsuragi commented:
Thank you for checking!

> After applying the changes you made to postgres_fdw_plus, I ran the regression test and unfortunately, it failed.

My PostgreSQL version was a little bit old. I got the same error with this commit (https://github.com/postgres/postgres/commit/e056c557aef4006c3dfbf8a4b94b7ae88eb9fd67).
Now, because the commit is reverted (https://github.com/postgres/postgres/commit/9ce04b50e120275afbc03ca0b80839dde3da8308), we do not get the same failure.

However, now I get another test failure by this commit (https://github.com/postgres/postgres/commit/6633cfb21691840c33816a6dacaca0b504efb895).
This commit affects main branch too, so I will created another PR to fix this.

MasaoFujii commented:
@YKatsuragi 
Thank you for reporting and diagnosing the issue! I have just pushed the fix to the main branch (https://github.com/pgfdwplus/postgres_fdw_plus/commit/46ff91847a6cbfa3fa392c3f423c5483d9123ea7), so we can restart the review of your patch to fix the parallel_commit=on issue as soon as possible.

MasaoFujii commented:
@YKatsuragi 

```
-                     application_name 'pgfdw_plus_loopback2'
+                     application_name 'pgfdw_plus_loopback2',
+					 parallel_commit 'on'
```

Could you please use spaces instead of tabs to indent the "parallel_commit" line?
This will help maintain consistency in the formatting of the SQL in regression test.

```
+	/*
+	 * If parallel_commit and pgfdw_skip_commit_phase are true, this connection
+	 * cache entry will not be used in this transaction.
+	 */
+	if (entry->parallel_commit && pgfdw_skip_commit_phase)
+	{
+		entry->fxid = InvalidFullTransactionId;
+		pgfdw_reset_xact_state(entry, true);
+	}
```

The same thing as above can be done as follows. Right? If yes, the following change looks simpler and more appropriate as the fix. Thought?

```
 				case XACT_EVENT_COMMIT:
 					pgfdw_commit_prepared(entry,
 										  &pending_entries_commit_prepared);
-					if (entry->parallel_commit)
+					if (entry->parallel_commit && !pgfdw_skip_commit_phase)
 						continue;
```

YKatsuragi commented:
Thank you for fixing and reviewing!

>  I have just pushed the fix to the main branch (46ff918)

I rebased to reflect this fix.

> ```
> -                     application_name 'pgfdw_plus_loopback2'
> +                     application_name 'pgfdw_plus_loopback2',
> +					 parallel_commit 'on'
> ```
> Could you please use spaces instead of tabs to indent the "parallel_commit" line?

Fixed to use spaces.

> The same thing as above can be done as follows. Right? If yes, the following change looks simpler and more appropriate as the fix. Thought?
> ```
>  				case XACT_EVENT_COMMIT:
>  					pgfdw_commit_prepared(entry,
>  										  &pending_entries_commit_prepared);
> -					if (entry->parallel_commit)
> +					if (entry->parallel_commit && !pgfdw_skip_commit_phase)
>  						continue;
> ```

I think yes, and this way is better. Fixed.
